backend/src/statistics/statistics.service.ts

```diff
@@ -0,0 +1,158 @@
+
+    return {
+      attemptSummary,
+      durationSummary,
+      peerSummary,
+      heatmapData,
+      networkData,
+    };
+  }
+
+  private async getAttemptSummaryByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: { users: { some: { id: userId } }, endTime: { not: null } },
+      include: {
+        question: {
+          select: { difficulty: true },
+        },
+      },
+    });
+    const map = {
+      [Difficulty.EASY]: 0,
+      [Difficulty.MEDIUM]: 0,
+      [Difficulty.HARD]: 0,
+    };
+    for (const session of data) {
+      const { difficulty } = session.question;
+      map[difficulty] = map[difficulty] + 1;
+    }
+    return map;
+  }
+
+  private async getDurationSummaryByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: {
+        users: { some: { id: userId } },
+        endTime: { not: null },
+        startTime: {
+          gte: subYears(new Date(), 1),
+        },
+      },
+      include: {
+        question: {
+          select: { difficulty: true },
+        },
+      },
+    });
+    return data
+      .filter((session) => !!session.endTime)
+      .map((session) => {
+        return {
+          difficulty: session.question.difficulty as Difficulty,
+          timetaken: differenceInSeconds(
+            session.endTime || new Date(), // Should be defined by selection.
+            session.startTime,
+          ),
+          date: session.startTime,
+        };
+      });
+  }
+
+  private async getPeerSummaryByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: { users: { some: { id: userId } }, endTime: { not: null } },
+      include: {
+        question: {
+          select: { title: true },
+        },
+        users: true,
+      },
+    });
+    return data.map((session) => {
+      return {
+        userName: session.users.find((user) => user.id !== userId)?.name || "",
+        questionTitle: session.question.title,
+        date: session.startTime,
+      };
+    });
+  }
+
+  private async getHeatmapDataByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: {
+        users: { some: { id: userId } },
+        endTime: { not: null },
+        startTime: {
+          gte: subYears(new Date(), 1),
+        },
+      },
+    });
+    return data.map((session) => {
+      return { date: session.startTime };
+    });
+  }
+
+  private async getNetworkDataByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: {
+        users: { some: { id: userId } },
+        endTime: { not: null },
+      },
+      include: {
+        question: {
+          include: {
+            topics: true,
+          },
+        },
+      },
+    });
+    const topics = new Map<number, string>();
+    const relations = new Map<number, number>();
+    for (const session of data) {
```

Comment:
I will review this later when i wake up. Brain ded at 2am so can't seem to understand what is going on :(